cmd/cert.go

```diff
@@ -0,0 +1,122 @@
+
+	return cli.Command{
+		Name:   "cert-rotate",
+		Usage:  "rotate all certificates without ca",
+		Action: clusterRotateFromCli,
+		Flags:  certFlags,
+	}
+}
+
+func ClusterRotate(
+	ctx context.Context,
+	rkeConfig *v3.RancherKubernetesEngineConfig,
+	dockerDialerFactory, localConnDialerFactory hosts.DialerFactory,
+	k8sWrapTransport k8s.WrapTransport, configDir string) (string, string, string, string, map[string]pki.CertificatePKI, error) {
+
+	log.Infof(ctx, "Rotating Kubernetes cluster certificates")
+	var APIURL, caCrt, clientCert, clientKey string
+	kubeCluster, err := cluster.ParseCluster(ctx, rkeConfig, clusterFilePath, configDir, dockerDialerFactory, localConnDialerFactory, k8sWrapTransport)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	err = kubeCluster.TunnelHosts(ctx, false)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	currentCluster, err := kubeCluster.GetClusterState(ctx)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	if currentCluster == nil {
+		log.Infof(ctx, "This is a newely built cluster, can't rotate certificates")
+		return APIURL, caCrt, clientCert, clientKey, nil, nil
+	}
+
+	err = cluster.RotateRKECertificates(ctx, kubeCluster, currentCluster)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	err = kubeCluster.SetUpHosts(ctx, true)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	if err := services.RestartControlPlane(ctx, kubeCluster.ControlPlaneHosts); err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	allHosts := hosts.GetUniqueHostList(kubeCluster.EtcdHosts, kubeCluster.ControlPlaneHosts, kubeCluster.WorkerHosts)
+	if err := services.RestartWorkerPlane(ctx, allHosts); err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	err = kubeCluster.SaveClusterState(ctx, &kubeCluster.RancherKubernetesEngineConfig)
+	if err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	APIURL = fmt.Sprintf("https://" + kubeCluster.ControlPlaneHosts[0].Address + ":6443")
+	clientCert = string(cert.EncodeCertPEM(kubeCluster.Certificates[pki.KubeAdminCertName].Certificate))
+	clientKey = string(cert.EncodePrivateKeyPEM(kubeCluster.Certificates[pki.KubeAdminCertName].Key))
+	caCrt = string(cert.EncodeCertPEM(kubeCluster.Certificates[pki.CACertName].Certificate))
+
+	if err := checkAllIncluded(kubeCluster); err != nil {
+		return APIURL, caCrt, clientCert, clientKey, nil, err
+	}
+
+	log.Infof(ctx, "Finished Rotating Kubernetes cluster certificates successfully")
+	return APIURL, caCrt, clientCert, clientKey, kubeCluster.Certificates, nil
+}
+
+func clusterRotateFromCli(ctx *cli.Context) error {
+	clusterFile, filePath, err := resolveClusterFile(ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to resolve cluster file: %v", err)
+	}
+	clusterFilePath = filePath
```

Comment:
how exactly the clusterFile parameter being used? Is it being passed as a reference somewhere? As in this particular func, I don't see filePath or clusterFilePath being used.

Reply:
clusterFilePath is a global variable we defined in cmd/up.go, and is being used to resolve cli file path, we refactored that in 0.2